src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+            securitySessionAuthDispatcher, /*wasChannelThrottled*/ false, sessionIdleManager);
+            var channelDispatcher = channelHandler.GetDispatcher();
+            outerChannel.ChannelDispatcher = securityChannel.ChannelDispatcher = channelDispatcher;
+            await channelHandler.OpenAsync();
+            return channelDispatcher;
+        }
+
+        private IChannel GetInnerChannel(IChannel outerChannel)
+        {
+            IChannel securityChannel = null;
+            /* TODO once we add more features
+            if (outerChannel is IInputChannel)
+            {
+                securityChannel = new SecurityInputChannel(listener, (IInputChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+            }
+            else if (outerChannel is IInputSessionChannel))
+            {
+                securityChannel = new SecurityInputSessionChannel(listener, (IInputSessionChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+            }
+            else if (outerChannel is IDuplexChannel))
+            {
+                securityChannel = new SecurityDuplexChannel(listener, (IDuplexChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+            }
+            else if (outerChannel is IDuplexSessionChannel))
+            {
+                securityChannel = new SecurityDuplexSessionChannel(listener, (IDuplexSessionChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+            }
+            else*/
+            if (outerChannel is IReplyChannel)
+            {
+                securityChannel = new SecurityReplyChannel(this, (IReplyChannel)outerChannel);
+            }
+            /* else if (listener.SupportsRequestReply && typeof(TChannel) == typeof(IReplySessionChannel))
+             {
+                 securityChannel = new SecurityReplySessionChannel(listener, (IReplySessionChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+             }
+             else
+             {
+                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.GetString(SR.UnsupportedChannelInterfaceType, typeof(TChannel))));
+             }*/
+
+             return (IChannel)securityChannel;
+        }
+
+    }
+
+   abstract class ServerSecurityChannel<UChannel>  where UChannel : class, IChannel
+    {
+        static MessageFault secureConversationCloseNotSupportedFault;
+        string secureConversationCloseAction;
+        SecurityListenerSettingsLifetimeManager settingsLifetimeManager;
+        bool hasSecurityStateReference;
+
+        internal SecurityProtocol SecurityProtocol { get;  set; }
+        public IReplyChannel OuterChannel { get; private set; }
+
+        protected ServerSecurityChannel( IReplyChannel outerChannel)
+            //: base(channelManager, innerChannel, securityProtocol) // not sure use of channel manager at this time
```

Comment:
This should be typed as UChannel. Any channel shape specific logic should be in the concrete class type.